lib/bootsnap/load_path_cache/core_ext/kernel_require.rb

```diff
@@ -13,72 +13,106 @@ def self.make_load_error(path)
 module Kernel
   alias_method :require_without_cache, :require
   def require(path)
+    return false if Bootsnap::LoadPathCache.loaded_features_index.key?(path)
     if resolved = Bootsnap::LoadPathCache.load_path_cache.find(path)
-      require_without_cache(resolved)
+      Bootsnap::LoadPathCache.load_path_cache.register(path, resolved) do
+        require_without_cache(resolved)
+      end
     else
       raise Bootsnap::LoadPathCache::CoreExt.make_load_error(path)
     end
   rescue Bootsnap::LoadPathCache::ReturnFalse
     return false
   rescue Bootsnap::LoadPathCache::FallbackScan
-    require_without_cache(path)
+    Bootsnap::LoadPathCache.load_path_cache.register(path) do
+      require_without_cache(path)
+    end
   end
 
   alias_method :load_without_cache, :load
   def load(path, wrap = false)
+    return false if Bootsnap::LoadPathCache.loaded_features_index.key?(path)
+
     if resolved = Bootsnap::LoadPathCache.load_path_cache.find(path)
-      load_without_cache(resolved, wrap)
+      Bootsnap::LoadPathCache.load_path_cache.register(path, resolved) do
+        load_without_cache(resolved, wrap)
+      end
     else
       # load also allows relative paths from pwd even when not in $:
       relative = File.expand_path(path)
       if File.exist?(File.expand_path(path))
-        return load_without_cache(relative, wrap)
+        return Bootsnap::LoadPathCache.load_path_cache.register(path) do
+          load_without_cache(relative, wrap)
+        end
       end
       raise Bootsnap::LoadPathCache::CoreExt.make_load_error(path)
     end
   rescue Bootsnap::LoadPathCache::ReturnFalse
     return false
   rescue Bootsnap::LoadPathCache::FallbackScan
-    load_without_cache(path, wrap)
+    Bootsnap::LoadPathCache.load_path_cache.register(path) do
+      load_without_cache(path, wrap)
+    end
   end
 end
 
 class << Kernel
   alias_method :require_without_cache, :require
   def require(path)
+    return false if Bootsnap::LoadPathCache.loaded_features_index.key?(path)
+
     if resolved = Bootsnap::LoadPathCache.load_path_cache.find(path)
-      require_without_cache(resolved)
+      Bootsnap::LoadPathCache.load_path_cache.register(path, resolved) do
+        require_without_cache(resolved)
```

Comment:
Would it make sense to make `require_without_cache` do the `Bootsnap::LoadPathCache.load_path_cache.register` and return false if loaded? Might simplify the logic and reduce the chance of calling `require_without_cache` improperly

Reply:
Yeah, I'll give that a spin.